internal/graph/check.go

```diff
@@ -984,6 +991,494 @@ func trySendUsersetsAndDeleteFromMap(ctx context.Context, usersetsMap usersetsMa
+		}
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, nil
+		case msg, ok := <-checkOutcomeChan:
+			if !ok {
+				return &ResolveCheckResponse{
+					Allowed: false,
+					ResolutionMetadata: &ResolveCheckResponseMetadata{
+						DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + commonParameters.dsCount.Load(),
+					},
+				}, nil
+			}
+			if msg.err != nil && !errors.Is(msg.err, context.Canceled) && !errors.Is(msg.err, context.DeadlineExceeded) {
+				return nil, msg.err
+			}
+			if msg.resp != nil && msg.resp.GetAllowed() {
+				return msg.resp, nil
+			}
+		}
+	}
+}
+
+func recursiveMatchUserUserset(ctx context.Context,
+	req *ResolveCheckRequest,
+	commonParameters *recursiveMatchUserUsersetCommonData) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "recursiveMatchUserUserset")
+	defer span.End()
+
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+
+	req.GetRequestMetadata().Depth--
+	if req.GetRequestMetadata().Depth == 0 {
+		return nil, ErrResolutionDepthExceeded
+	}
+
+	commonParameters.dsCount.Add(1)
+
+	storeID := req.GetStoreID()
+	reqTupleKey := req.GetTupleKey()
+
+	opts := storage.ReadUsersetTuplesOptions{
+		Consistency: storage.ConsistencyOptions{
+			Preference: req.GetConsistency(),
+		},
+	}
+	iter, err := commonParameters.ds.ReadUsersetTuples(ctx, storeID, storage.ReadUsersetTuplesFilter{
+		Object:                      reqTupleKey.GetObject(),
+		Relation:                    reqTupleKey.GetRelation(),
+		AllowedUserTypeRestrictions: commonParameters.allowedUserTypeRestrictions,
+	}, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	filteredIter := storage.NewConditionsFilteredTupleKeyIterator(
+		storage.NewFilteredTupleKeyIterator(
+			storage.NewTupleKeyIteratorFromTupleIterator(iter),
+			validation.FilterInvalidTuples(commonParameters.typesys),
+		),
+		checkutil.BuildTupleKeyConditionFilter(ctx, req.GetContext(), commonParameters.typesys),
+	)
+	defer filteredIter.Stop()
+
+	// First, we will want to check if there are any intersection with the userToUsersetMapping.
+	// If so, we do not need to recursively lookup further.
+	var usersetItems []string
+	for {
+		t, err := filteredIter.Next(ctx)
+		if err != nil {
+			if errors.Is(err, storage.ErrIteratorDone) {
+				break
+			}
+			return nil, err
+		}
+		usersetName, relation := tuple.SplitObjectRelation(t.GetUser())
+		if relation == "" {
+			// This should not be possible given we have put in the type restriction.
+			return nil, fmt.Errorf("expect user %s to have form objection relation", t.GetUser())
+		}
+		if commonParameters.userToUsersetMapping.Exists(usersetName) {
+			return &ResolveCheckResponse{
+				Allowed: true,
+				ResolutionMetadata: &ResolveCheckResponseMetadata{
+					DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + commonParameters.dsCount.Load(),
+				},
+			}, nil
+		}
+		usersetItems = append(usersetItems, usersetName)
+	}
```

Comment:
Could we re-use what we already have? (Disclaimer: haven't tested)

```suggestion
objectType := tuple.GetType(req.GetTupleKey().GetObject())
relation := req.GetTupleKey().GetRelation()

cancellableCtx, cancelFunc := context.WithCancel(ctx)
defer cancelFunc()

objectToUsersetMessageChan := streamedLookupUsersetForObject(cancellableCtx, commonParameters.typesys, commonParameters.ds, req, []*openfgav1.RelationReference{
	typesystem.DirectRelationReference(objectType, relation),
}, commonParameters.maxConcurrentReads)

var usersetItems []string
for usersetMsg := range objectToUsersetMessageChan {
	if usersetMsg.err != nil {
		return nil, usersetMsg.err
	}
	usersetName := usersetMsg.userset
	if commonParameters.userToUsersetMapping.Exists(usersetName) {
		return &ResolveCheckResponse{
			Allowed: true,
			ResolutionMetadata: &ResolveCheckResponseMetadata{
				DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + commonParameters.dsCount.Load(),
			},
		}, nil
	}
	usersetItems = append(usersetItems, usersetName)
}
```

Reply:
We can.  However, calling streamedLookupUsersetForObject will create un-necessary go-routine since we do not have anything to parallelize (i.e., this is a single producer single consumer).

Reply:
Will note in comment.

Reply:
True, but the tradeoff is that you reduce code duplication a lot, which I think makes it worth it because this code will then be more easily adaptable to handle the nested TTU scenario.

Reply:
Updated